auth/services/util.go

```diff
@@ -1,22 +1,87 @@
 package services
 
 import (
+	"fmt"
 	ssi "github.com/nuts-foundation/go-did"
 	"github.com/nuts-foundation/go-did/did"
 	"github.com/nuts-foundation/nuts-node/vdr/types"
 	"time"
 )
 
-// ResolveEndpointURL finds the endpoint with the given type of the given holder and unmarshals it as single URL.
+// ResolveServiceURL finds the endpoint with the given type of the given holder referenced from the given service, and unmarshals it as single URL.
 // It returns the endpoint ID and URL, or an error if anything went wrong;
 // - holder document can't be resolved,
 // - service with given type doesn't exist,
 // - multiple services match,
-// - serviceEndpoint isn't a string.
-func ResolveEndpointURL(resolver types.DocResolver, holder did.DID, endpointType string, validAt *time.Time) (endpointID ssi.URI, endpointURL string, err error) {
+// - referenced service doesn't resolve.
+// - referenced service not unique.
+func ResolveServiceURL(resolver types.DocResolver, holder did.DID, serviceType string, endpointType string, validAt *time.Time) (ssi.URI, string, error) {
 	doc, _, err := resolver.Resolve(holder, &types.ResolveMetadata{ResolveTime: validAt})
 	if err != nil {
 		return ssi.URI{}, "", err
 	}
-	return doc.ResolveEndpointURL(endpointType)
+
+	compoundService, err := resolveCompoundService(*doc, serviceType)
+	if err != nil {
+		return ssi.URI{}, "", fmt.Errorf("service not found (did=%s, service=%s)", doc.ID, serviceType)
+	}
+
+	endpointReference, ok := compoundService[endpointType]
+	if !ok {
+		return ssi.URI{}, "", fmt.Errorf("endpointType for service not found (did=%s, service=%s, type=%s)", doc.ID, serviceType, endpointType)
+	}
+	refDID, err := ssi.ParseURI(endpointReference)
+	if err != nil {
+		return ssi.URI{}, "", err
+	}
+
+	uri, url, err := resolveEndpointURL(resolver, *refDID, validAt)
+	if err != nil {
+		return ssi.URI{}, "", fmt.Errorf("failed to resolve endpoint from service (did=%s, service=%s, type=%s)", doc.ID, endpointReference, endpointType)
+	}
+
+	return uri, url, nil
+}
+
+type compoundServiceType map[string]string
+
+func resolveCompoundService(doc did.Document, serviceType string) (service compoundServiceType, err error) {
+	cs := compoundServiceType{}
+	var services []did.Service
+	for _, service := range doc.Service {
+		if service.Type == serviceType {
+			services = append(services, service)
+		}
+	}
+	if len(services) != 1 {
+		return cs, fmt.Errorf("incorrect number of services (did=%s, service=%s)", doc.ID, serviceType)
```

Comment:
'incorrect number of services' doesn't tell the operator what was expected and/or how many were found. I'd at least rewrite it to; "expected 1 service" (and maybe add; "found %d")